codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -137,145 +207,111 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
-                int streamId = entry.key();
-                if (streamId > goAwayFrame.lastStreamId() && isOutboundStream(server, streamId)) {
-                    childChannel.pipeline().fireUserEventTriggered(goAwayFrame.retainedDuplicate());
-                }
+            final Http2GoAwayFrame goAwayFrame = (Http2GoAwayFrame) msg;
+            try {
+                forEachActiveStream(new Http2FrameStreamVisitor() {
+                    @Override
+                    public boolean visit(Http2FrameStream stream) {
+                        final int streamId = stream.id();
+                        final DefaultHttp2StreamChannel childChannel = channels.get(stream);
+                        if (streamId > goAwayFrame.lastStreamId() && isOutboundStream(server, streamId)) {
+                            childChannel.pipeline().fireUserEventTriggered(goAwayFrame.retainedDuplicate());
+                        }
+                        return true;
+                    }
+                });
+            } finally {
+                // We need to ensure we release the goAwayFrame.
+                goAwayFrame.release();
+            }
+        } else if (msg instanceof Http2SettingsFrame) {
+            Http2Settings settings = ((Http2SettingsFrame) msg).settings();
+            if (settings.initialWindowSize() != null) {
+                initialOutboundStreamWindow = settings.initialWindowSize();
             }
-            goAwayFrame.release();
-        } else {
-            // It's safe to release, as UnsupportedMessageTypeException just calls msg.getClass()
-            ReferenceCountUtil.release(msg);
-            throw new UnsupportedMessageTypeException(msg);
         }
     }
 
-    private void fireChildReadAndRegister(Http2StreamChannel childChannel, Http2StreamFrame frame) {
-        // Can't use childChannel.fireChannelRead() as it would fire independent of whether
-        // channel.read() had been called.
-        childChannel.fireChildRead(frame);
-        if (!childChannel.inStreamsToFireChildReadComplete) {
-            channelsToFireChildReadComplete.add(childChannel);
-            childChannel.inStreamsToFireChildReadComplete = true;
-        }
+    private void channelReadStreamFrame(Http2StreamFrame frame) {
+        Http2FrameStream stream = frame.stream();
+
+        fireChildReadAndRegister(channels.get(stream), frame);
     }
 
-    @Override
-    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
-        if (evt instanceof Http2StreamActiveEvent) {
-            Http2StreamActiveEvent activeEvent = (Http2StreamActiveEvent) evt;
-            onStreamActive(activeEvent.streamId(), activeEvent.headers());
-        } else if (evt instanceof Http2StreamClosedEvent) {
-            onStreamClosed(((Http2StreamClosedEvent) evt).streamId());
-        } else {
-            ctx.fireUserEventTriggered(evt);
+    private void onStreamClosed(Http2FrameStream stream) {
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)